stdlib/format.ml

```diff
@@ -384,30 +420,37 @@ let format_pp_token state size = function
     if state.pp_current_indent != state.pp_margin - state.pp_space_left
     then pp_skip_token state
 
+  | Pp_string_if_newline s ->
+    if state.pp_is_new_line
+    then format_pp_text state (String.length s) s
+
   | Pp_break { fits; breaks } ->
-    let before, off, _ = breaks in
+    format_pp_break state size fits breaks
+
+  | Pp_break_or_string_if_newline (n, off, s) ->
+    if state.pp_is_new_line
+    then format_pp_text state size s
+    else
+      let fits = ("", n, "") and breaks = ("", off, "") in
+      format_pp_break state size fits breaks
+
+  | Pp_fits_or_breaks (fits, n, off, breaks) ->
     begin match Stack.top_opt state.pp_format_stack with
     | None -> () (* No open box. *)
-    | Some { box_type; width } ->
-      begin match box_type with
-      | Pp_hovbox ->
-        if size + String.length before > state.pp_space_left
-        then break_new_line state breaks width
-        else break_same_line state fits
-      | Pp_box ->
-        (* Have the line just been broken here ? *)
-        if state.pp_is_new_line then break_same_line state fits else
-        if size + String.length before > state.pp_space_left
-          then break_new_line state breaks width else
-        (* break the line here leads to new indentation ? *)
-        if state.pp_current_indent > state.pp_margin - width + off
-        then break_new_line state breaks width
-        else break_same_line state fits
-      | Pp_hvbox -> break_new_line state breaks width
-      | Pp_fits -> break_same_line state fits
-      | Pp_vbox -> break_new_line state breaks width
-      | Pp_hbox -> break_same_line state fits
-      end
+    | Some { box_type= ty; width } ->
+        let text =
+          if ty = Pp_fits then
+            fits
+          else begin
+            if off >= 0 then begin
+              if size + String.length breaks >= state.pp_space_left
+              then break_new_line state ("", off, "") width
+              else break_same_line state ("", n, "")
+            end;
+            breaks
+          end in
+        if text <> ""
+        then format_pp_text state size text
```

Comment:
I find it a bit strange that `size` is used here, even in the case where the line has been broken. In that case, given the rest of the interpreter code, I would expect `String.length text` to be used; but then I don't really understand what the interpreter code is doing.

I wonder if `n` and `off` could be replaced by a full `{ fits; breaks }` record as mentioned above. Finally, the `off >= 0` conditional is also a bit strange, in the sense that it is consistent with anything that currently exists in the implementation. (What happens if that test is removed? Do the examples behave incorrectly? Are we encoding the more structured idea of having an optional break here?)

Reply:
It was strange indeed, I now use `format_string` that correctly use the length of the string. I didn't rewrite this Pp_fits_or_breaks case using `{ fits; breaks }` because the strings `fits` and `breaks` are not actually used when calling `break_same/new_line` so we can't reuse `pp_print_custom_break` or have something similar.
I removed the `off >= 0` condition.